File: language.cpp
#include <cstdlib>
#include <iostream>
#include <string>
#include <fstream>
#include <filesystem>
#include <vector>
#include <utility>

#include <unistd.h>

using namespace std;

enum class System{
    Ubuntu,
    Debian,
    ArchLinux,
    Unsupported };
enum class User{
    Root,
    Other };

auto CheckIfRunningInsideWSL(){
    if (filesystem::exists("/proc/version")) {
        ifstream version_file("/proc/version");
        string version_info = "";
        getline(version_file, version_info);
        if (version_info.find("Microsoft") != string::npos or 
        version_info.find("WSL") != string::npos) {
            return true; }
        version_file.close(); }
    if (filesystem::exists("/proc/sys/kernel/osrelease")) {
        ifstream osrelease_file("/proc/sys/kernel/osrelease");
        string osrelease_info = "";
        getline(osrelease_file, osrelease_info);
        if (osrelease_info.find("microsoft") != string::npos or
            osrelease_info.find("WSL") != string::npos) {
            return true; }
        osrelease_file.close(); }
    return false; }

auto GetCurrentUser(){
    if (getenv("USER") != nullptr){
        string user = getenv("USER");
        if (user == "root" or geteuid() == 0){
            return User::Root; }
        else{
            return User::Other; } }
    else{
        throw runtime_error("无法获取用户信息"); } }

/// @brief  获取当前系统的发行版名称
/// @return 发行版名称，带有版本号
/// @author ChatGPT
auto GetCurrentSystem() {
    string distro_name = "";
    ifstream release_file("/etc/os-release");
    if (release_file.is_open()) {
        string line = "";
        while (getline(release_file, line)) {
            if (line.find("PRETTY_NAME=") != string::npos) {
                distro_name = line.substr(line.find("=") + 1);
                distro_name.erase(0, 1); // remove opening quote
                distro_name.erase(distro_name.size() - 1); // remove closing quote
                break; } }
        release_file.close(); }

    if(distro_name.find("Ubuntu") != string::npos){
        return pair(System::Ubuntu, distro_name); }
    else if(distro_name.find("Debian") != string::npos or
            distro_name.find("Armbian") != string::npos){
        return pair(System::Debian, distro_name); }
    else if(distro_name.find("Arch Linux") != string::npos or
            distro_name.find("CachyOS") != string::npos){
        return pair(System::ArchLinux, distro_name); }
    else{
        return pair(System::Unsupported, distro_name); } }

auto check() {
    cout << "请输入您的选项（输入y或者yes继续）";
    string userin = "";
    cin >> userin;
    if (userin == "y" or userin == "yes")
        return;
    else{
        cout << "输入不正确欸，如果想停止本程序的话请按下Ctrl+C" << endl;
        check(); } }

auto Ubuntu(){
    cout << "即将开始安装并配置简体中文语言包，请输入y或者yes继续，或者按Ctrl+C终止本程序";
    check();
    cout << "开始更新语言配置" << endl;
    system("apt update");
    system("apt install -y language-pack-zh-hans");
    system("update-locale LANG=zh_CN.UTF-8");
    cout << "配置完成，更改将在您下一次登录Shell时应用" << endl; }

auto Debian(){
    cout << "即将开始更新语言配置，请输入y或者yes继续，或者按Ctrl+C终止本程序";
    check();
    cout << "开始更新语言配置" << endl;
    fstream locale_gen("/etc/locale.gen");
    if (locale_gen.is_open() == false){
        throw system_error(errno, system_category(), "未能正确打开/etc/locale.gen"); }
    vector<string> FileContent;
    string Buffer = "";
    while (getline(locale_gen, Buffer)){
        FileContent.push_back(Buffer); }
    for (auto& Line : FileContent){
        if (Line == "zh_CN.UTF-8 UTF-8"){
            goto skip_write; } }
    locale_gen.clear();
    locale_gen.seekp(0, ios::end);
    locale_gen << '\n' << "zh_CN.UTF-8 UTF-8" << endl;
skip_write:
    locale_gen.close();
    system("locale-gen");
    system("update-locale LANG=zh_CN.UTF-8");
    cout << "配置完成，更改将在您下一次登录Shell时应用" << endl; }

auto ArchLinux(){
    // https://wiki.archlinuxcn.org/wiki/Locale
    cout << "即将开始更新语言配置，请输入y或者yes继续，或者按Ctrl+C终止本程序";
    check();
    cout << "开始更新语言配置" << endl;
    system("localectl set-locale LANG=zh_CN.UTF-8");
    if (CheckIfRunningInsideWSL() == true){
        // https://wiki.archlinuxcn.org/wiki/%E5%9C%A8_WSL_%E4%B8%8A%E5%AE%89%E8%A3%85_Arch_Linux#%E4%BF%AE%E6%94%B9%E5%8C%BA%E5%9F%9F%E8%AE%BE%E7%BD%AE
        cout << "检测到您正在WSL中运行Arch Linux，正在创建 /etc/locale.conf -> /etc/default/locale 符号链接" << endl;
        filesystem::create_symlink("/etc/locale.conf", "/etc/default/locale"); }
    cout << "配置完成，更改将在您下一次登录Shell时应用" << endl; }

auto ThrowUnsupportedSystemError(){
    cout << "错误：未知操作系统" << endl
        << "当前仅适配了Ubuntu，Debian和Arch Linux" << endl
        << "强行使用可能会造成不可预知的问题" << endl
        << "为避免出现故障，程序会强制退出" << endl
        << "如果是误报，请在GitHub仓库打开一个新的issue。" << endl;
    throw runtime_error("不支持的操作系统"); }

int main() {
    auto OsType = GetCurrentSystem();
    if (OsType.first == System::Unsupported){
        ThrowUnsupportedSystemError(); }

    if (GetCurrentUser() == User::Other){
        cout << "更改语言文件需要root权限，请使用sudo重新运行本程序" << endl;
        throw runtime_error("权限不足"); }

    cout << "检测到您使用的系统为：" << OsType.second << endl;
    if (OsType.first == System::Ubuntu){ Ubuntu(); }
    if (OsType.first == System::Debian){ Debian(); }
    if (OsType.first == System::ArchLinux){ ArchLinux(); } }

File: language.cs
#!/usr/bin/env -S dotnet --
// https://learn.microsoft.com/zh-cn/dotnet/core/sdk/file-based-apps

#:property PublishAot=true
#:property StaticExecutable=true
#:property OptimizationPreference=Speed
// #:property OptimizationPreference=Size
#:property InvariantGlobalization=true
// ↑ 这个东西需要libicu
// #:property StackTraceSupport=false

using System.Diagnostics;
using System.Runtime.InteropServices;
using System.Text;

void CheckUser(string DisplayMessage = "即将开始更新语言配置"){
    Console.WriteLine(DisplayMessage);
    Console.WriteLine("按回车继续，按Esc取消");
    while (true){
        var KeyInfo = Console.ReadKey(true);
        if (KeyInfo.Key == ConsoleKey.Enter){
            return; }
        else if (KeyInfo.Key == ConsoleKey.Escape){
            Environment.Exit(0); } } }

bool IsRunningInsideWSL(){
    string VersionFile = File.ReadAllText("/proc/version");
    string KernelOsReleaseFile = File.ReadAllText("/proc/sys/kernel/osrelease");
    if (VersionFile.Contains("Microsoft") || VersionFile.Contains("WSL")){
        return true; }
    if (KernelOsReleaseFile.Contains("microsoft") || KernelOsReleaseFile.Contains("WSL")){
        return true; }
    return false; }

void ExecuteCommand(string Command, 
                    List<string>? Arguments = null,
                    Action<int, string, string?>? WhenErrorThrown = null){
    List<string> StandardError = [];
    int ExitCode = 0;

    try{
        using var Process = new Process();
        Process.StartInfo.FileName = Command;
        if (Arguments != null){
            foreach (var Argument in Arguments){
                Process.StartInfo.ArgumentList.Add(Argument); } }
        Process.StartInfo.RedirectStandardError = true;
        Process.ErrorDataReceived += (_, Event) => {
            if (Event.Data != null){
                Console.WriteLine(Event.Data);
                StandardError.Add(Event.Data + "\n"); } };
        Process.Start();
        Process.BeginErrorReadLine();
        Process.WaitForExit();
        ExitCode = Process.ExitCode;
        if (ExitCode != 0){
            throw new Exception($"命令执行失败，退出代码: {ExitCode}"); } }
    catch (Exception ExceptionInstance){
        WhenErrorThrown ??= (ExitCode, StandardError, ExceptionMessage) => {
            Console.WriteLine(@$"
            命令执行失败
            退出代码: {ExitCode}
            stderr: {StandardError}
            异常信息: {ExceptionMessage ?? ""}"); };
        WhenErrorThrown(ExitCode, string.Join("\n", StandardError), ExceptionInstance.Message);
        throw; } }

string OsDescription = RuntimeInformation.OSDescription;
if (Enum.TryParse<SupportedLinuxDistributions>(OsDescription.Split(' ')[0], out var CurrentSystem) == false){
    if (OsDescription.Contains("Arch Linux")){
        CurrentSystem = SupportedLinuxDistributions.ArchLinux; }
    else if (OsDescription.Contains("Rocky Linux")){
        CurrentSystem = SupportedLinuxDistributions.RockyLinux; }
    else{
        string CurrentSupportedLinuxDistributions = string.Join(", ", Enum.GetNames<SupportedLinuxDistributions>());
        string Message = @$"
            当前系统为 {OsDescription}
            当前仅支持 {CurrentSupportedLinuxDistributions}
            强行使用可能会造成不可预知的问题
            为避免出现故障，程序会强制退出
            如果是误报，请在GitHub仓库打开一个新的issue。
        ".Trim();
        throw new NotSupportedException(Message); } }
if (Environment.IsPrivilegedProcess == false) {
    using var Process = new Process();
    Process.StartInfo.FileName = "sudo";
    Process.StartInfo.ArgumentList.Add(Environment.ProcessPath ?? throw new Exception("无法获取当前程序路径"));
    Console.WriteLine("更改语言文件需要root权限，当前用户权限不足");
    Console.WriteLine("正在尝试使用sudo重新运行程序");
    Console.WriteLine("如果您当前登录的用户拥有密码，请在sudo提示符中输入密码");
    Process.Start();
    Process.WaitForExit();
    Environment.Exit(Process.ExitCode); }

Console.WriteLine($"探测到当前系统为 {OsDescription}");
switch (CurrentSystem){
    case SupportedLinuxDistributions.Debian or SupportedLinuxDistributions.Armbian: {
        CheckUser();
        Console.WriteLine("正在检查语言配置");
        string[] NewContent = [];
        using var FileStream = new FileStream("/etc/locale.gen", FileMode.Open, FileAccess.ReadWrite, FileShare.None);
        using (var StreamReader = new StreamReader(FileStream, Encoding.UTF8, true, 1024, true)){
            NewContent = [.. StreamReader.ReadToEnd()
            .Split("\n")
            .Select(Line => Line.Trim())
            .Select(Line => {
                if (Line.Contains("zh_CN.UTF-8 UTF-8")){
                    Line = "zh_CN.UTF-8 UTF-8"; }
                return Line; } ) ]; }
        FileStream.SetLength(0);
        FileStream.Seek(0, SeekOrigin.Begin);
        using (var StreamWriter = new StreamWriter(FileStream, Encoding.UTF8)){
            foreach (var Line in NewContent){
                StreamWriter.WriteLine(Line); } }
        Console.WriteLine("正在生成语言配置");
        ExecuteCommand("locale-gen");
        Console.WriteLine("正在更新语言配置");
        ExecuteCommand("update-locale", ["LANG=zh_CN.UTF-8"]);
        break; }
    case SupportedLinuxDistributions.Ubuntu: {
        CheckUser("即将开始安装中文语言包");
        Console.WriteLine("正在更新apt软件包源");
        ExecuteCommand("apt", ["update"]);
        Console.WriteLine("正在安装中文语言包");
        ExecuteCommand("apt", ["install", "language-pack-zh-hans", "-y"]);
        Console.WriteLine("正在更新语言配置");
        ExecuteCommand("update-locale", ["LANG=zh_CN.UTF-8"]);
        break; }
    case SupportedLinuxDistributions.ArchLinux or SupportedLinuxDistributions.CachyOS: {
        CheckUser();
        Console.WriteLine("开始更新语言配置");
        ExecuteCommand("localectl", ["set-locale", "LANG=zh_CN.UTF-8"]);
        if (IsRunningInsideWSL()){
            // https://wiki.archlinuxcn.org/wiki/%E5%9C%A8_WSL_%E4%B8%8A%E5%AE%89%E8%A3%85_Arch_Linux#%E4%BF%AE%E6%94%B9%E5%8C%BA%E5%9F%9F%E8%AE%BE%E7%BD%AE
            Console.WriteLine("检测到您正在WSL中运行Arch Linux，正在创建 /etc/locale.conf -> /etc/default/locale 符号链接");
            if (File.Exists("/etc/default/locale")){
                File.Move("/etc/default/locale", "/etc/default/locale.bak", true); }
            File.CreateSymbolicLink("/etc/default/locale", "/etc/locale.conf"); }
        break; }
    case SupportedLinuxDistributions.RockyLinux or SupportedLinuxDistributions.AlmaLinux or SupportedLinuxDistributions.Fedora: {
        CheckUser("即将开始安装中文语言包");
        Console.WriteLine("开始更新语言配置");
        Console.WriteLine("正在安装中文语言包");
        ExecuteCommand("dnf", ["install", "-y", "--nogpgcheck", "glibc-langpack-zh", "langpacks-zh_CN"]);
        Console.WriteLine("正在更新语言配置");
        ExecuteCommand("localectl", ["set-locale", "LANG=zh_CN.UTF-8"]);
        
        if (IsRunningInsideWSL() && CurrentSystem == SupportedLinuxDistributions.AlmaLinux) {
            Console.ForegroundColor = ConsoleColor.Yellow;
            Console.WriteLine();
            Console.WriteLine("警告：检测到您正在WSL内使用AlmaLinux");
            Console.WriteLine("目前已知AlmaLinux的WSL镜像可能存在一些问题");
            Console.WriteLine("如果您遇到了诸如 dnf nano 等软件的显示仍然为英文的问题");
            Console.WriteLine("请尝试使用dnf重新安装或升级出现问题的软件包，或者直接完整更新系统内所有软件包，这应当可以解决问题");
            Console.WriteLine();
            Console.ResetColor(); }
        break; }
    default: {
        Console.ForegroundColor = ConsoleColor.Red;
        Console.WriteLine($"错误：您当前使用的操作系统 {OsDescription} ,匹配名称 {CurrentSystem} 受到支持，但是lag2cn没有为其配置执行逻辑");
        Console.WriteLine("这是一个程序Bug，请前往Github仓库开启一个issue，并提供当前输出日志");
        Console.ResetColor();
        throw new InvalidOperationException("遗漏的switch匹配"); } }

Console.WriteLine("配置完成，更改将在您下一次登录Shell时生效，按任意键退出");
Console.ReadKey(true);

enum SupportedLinuxDistributions {
    Debian,
    Armbian,
    Ubuntu,
    ArchLinux,
    CachyOS,
    RockyLinux,
    AlmaLinux,
    Fedora }